prediction_market_agent/agents/arbitrage_agent/data_models.py

```diff
@@ -0,0 +1,93 @@
+import typing as t
+
+from prediction_market_agent_tooling.markets.agent_market import AgentMarket
+from prediction_market_agent_tooling.tools.betting_strategies.utils import SimpleBet
+from prediction_market_agent_tooling.tools.utils import check_not_none
+from pydantic import BaseModel
+
+
+class Correlation(BaseModel):
+    near_perfect_correlation: bool | None
+    reasoning: str
+
+
+class ArbitrageBet(BaseModel):
+    main_market_bet: SimpleBet
+    related_market_bet: SimpleBet
+
+
+class CorrelatedMarketPair(BaseModel):
+    main_market: AgentMarket
+    related_market: AgentMarket
+    correlation: Correlation
+
+    def __str__(self) -> str:
+        return f"main_market {self.main_market.question} related_market_question {self.related_market.question} potential_profit_per_unit {self.potential_profit_per_bet_unit()}"
+
+    def potential_profit_per_bet_unit(self) -> float:
+        """
+        Calculate potential profit per bet unit based on high positive market correlation.
+        For positively correlated markets: Bet YES/NO or NO/YES.
+        """
+
+        if self.correlation.near_perfect_correlation is None:
+            return 0
+
+        bet_direction_main, bet_direction_related = self.bet_directions()
+        p_main = (
+            self.main_market.current_p_yes
+            if bet_direction_main
+            else self.main_market.current_p_no
+        )
+        p_related = (
+            self.related_market.current_p_yes
+            if bet_direction_related
+            else self.related_market.current_p_no
+        )
+        denominator = p_main + p_related
+        return (1 / denominator) - 1
+
+    def bet_directions(self) -> t.Tuple[bool, bool]:
+        correlation = check_not_none(self.correlation.near_perfect_correlation)
+        if correlation:
+            # We compare denominators for cases YES/NO and NO/YES bets and take the most profitable (i.e. where denominator is the lowest).
+            # For other cases we employ similar logic.
+            yes_no = self.main_market.current_p_yes + self.related_market.current_p_no
+            no_yes = self.main_market.current_p_no + self.related_market.current_p_yes
+            return (True, False) if yes_no <= no_yes else (False, True)
+
+        else:
```

Comment:
Could you explain to me, why if `near_perfect_correlation == False`, why bet `Yes, Yes` or `No, No`? If it's not perfectly correlated (either positively or negatively), shouldn't it be skipped by this agent?

Reply:
Sure. In the case of -1 correlation, let's assume we have these markets:
-> Will Kamala be elected president?
-> Will Trump be elected president?

We should EITHER bet a. YES on both OR b. NO on both. Since correlation = -1, we will ALWAYS win exactly one of the bets.

If the above is correct, the question is only where is it "cheaper" to place the bet - if both markets are 50-50 then there is no difference, but for unbalanced markets there is a formula that dictates if bets should be (YES, YES) or (NO, NO). Maybe you can try deriving for yourself, if you have problems, happy to share my notes.